MooseVariable -> MooseObject

lindsayad opened the pull request:
Will allow:

- MooseVariable to have restartable data
- consistent interface/behavior with other MooseObjects
- cleaner code-base

aeslaughter committed:
WIP: MooseVariableBase -> MooseObject

aeslaughter committed:
Close, need to get order/family from MooseEnum

aeslaughter committed:
WIP

aeslaughter committed:
simple diffusion not crashing

lindsayad committed:
Compile with modern moose

lindsayad committed:
Simple diffusion runs

lindsayad committed:
Working on making moose tests pass

lindsayad committed:
Remove "_displaced" design

lindsayad committed:
It's a miracle. All tests passing in serial locally

Refs #10601

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/372839/) on 891e8b6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13769/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13769/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7639d7ade340b9bdde9532715a6cc2e703a1cbd4`

lindsayad committed:
Now lets make everything else pass

lindsayad commented:
@aeslaughter What would be your recommendation for resolving the Documentation failure? I'm thinking about making the `AddElementalFieldAction` triggered by indicator and marker syntax not inherit from `AddVariableAction` since I already have to do some workarounds to prevent the `type` of the Indicator or Marker from being used in the `AddElementalFieldAction`.

But that still leaves me with the fact that both `AddVariableAction` and `AddAuxVariableAction` both build MooseVariables, and I don't want to change that. Should I write documentation pages for the different MooseVariable types in `/Variables` and then just sym-link to those files from `/AuxVariables`?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/372981/) on 11b46b5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13769/site)

This comment will be updated on new commits.

lindsayad commented:
moose test timeout will be fixed by https://github.com/libMesh/libmesh/pull/2190

aeslaughter commented:
@lindsayad I would just symlink to the pages that are the same.

lindsayad commented:
> I would just symlink to the pages that are the same.

I like it!

lindsayad committed:
Documentation

aeslaughter commented on framework/doc/content/source/variables/MooseVariableBase.md:
````diff
@@ -0,0 +1,166 @@
+
+```
+[AuxVariables]
+  [./aux_scalar]
````
```suggestion
  [aux_scalar]
```

aeslaughter commented on framework/doc/content/source/variables/MooseVariableBase.md:
```diff
@@ -0,0 +1,166 @@
+  [./aux_scalar]
+    order = SECOND
+    family = SCALAR
+  [../]
```
```suggestion
  []
```

lindsayad committed:
Let AddVariableAction set family, order, scaling pars

This improves backwards compatability so that applications
which add variables simply by creating variable actions (e.g.
no input file Variables block) will still work

lindsayad committed:
Initialize _fe_type and _component in constructor

Marmot has an action which inherits from AddAuxVariableAction
and which makes use of the _fe_type and _scalar_var members
in its `act` override. For backwards compatibility, we can't
rely on the `act` override calling `init` so we have to make sure
that _fe_type and _scalar_var and init'd in the constructor.
Again we are making use of the fact that we've duplicated params
between the MooseVariable* MooseObject and the AddVariableAction
action object

lindsayad commented:
Finally no app failures. Woo!

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -916,6 +931,8 @@ class InputParameters : public Parameters
   // These are the only objects allowed to _create_ InputParameters
   friend InputParameters emptyInputParameters();
+  // friend InputParameters validParams<MooseApp>();
+  // friend class MooseVariableBase; //TODO:MooseVariableToMooseObject
```
they were left over from you haha. I'll remove them

gambka commented:
@lindsayad This causes a large amount of failures in modules and other applications when executed with the --error-deprecated parameter passed to ./run_tests related to an API change in the adding of Variables and AuxVariables.

I particularly dislike having the deprecated syntax in the code as new users may take that and continue to use it.  Should moose, the modules, and applications for which we test regularly pass the error-deprecated when making these sort of fundamental API changes? Then the deprecation error is there for untracked apps/inputs to give time for their transition?

lindsayad commented:
> This causes a large amount of failures in modules and other applications when executed with the --error-deprecated parameter passed to ./run_tests related to an API change in the adding of Variables and AuxVariables.

It seems that the most deprecation messages are about the switch from MeshModifiers from MeshGenerators...but yes I definitely see some from this PR as I would expect.

>  Should moose, the modules, and applications for which we test regularly pass the error-deprecated when making these sort of fundamental API changes? Then the deprecation error is there for untracked apps/inputs to give time for their transition?

We do deprecation testing all over CIVET, like [here](https://moosebuild.inl.gov/job/199724/) and [here](https://civet.inl.gov/job/382429/), which I'm sure you're aware of because that is probably what brought this to your attention. So what change are you proposing? That we make those oranges boxes be red? Is the hope then that the MOOSE developer be the one who goes through and updates all the applications to use the new API?

gambka commented:
I found it out because I was waiting for the MeshGenerators merge to go in to put a PR to fix the majority of deprecation errors (there is some we can't fix due to another issue #13841). It looks like more are coming from the MeshModifiers/nesting of MeshGenerators because the max_failures is limited to 50.  If that is increased, the majority of failures come from this because we have a lot of tests that use the actions (which appears to be what this API change is for).  

I think as a group we should try to keep the deprecated boxes from ever being orange (or as infrequently as possible). As at some point someone will take the deprecated approach as an example and use and there will be more work when we remove the support for the deprecated approach.  Sure CIVET does testing but if I wasn't waiting for the MeshGenerators I wouldn't have noticed these other failures.  Some people do not update their code that frequently to see the deprecation messages either. Perhaps an email to the mailing list when such a fundamental change to the API is changed? I am fine with making the necessary API changes to the downstream apps I use but as a group we should try to make all our source code and tests use non-deprecated versions of the code as soon as possible.

lindsayad commented:
> I think as a group we should try to keep the deprecated boxes from ever being orange (or as infrequently as possible).

I agree 100%.

> As at some point someone will take the deprecated approach as an example and use and there will be more work when we remove the support for the deprecated approach.

Absolutely a concern, although I know we MOOSE developers are actually notoriously bad at ever removing the deprecated API.

> Sure CIVET does testing but if I wasn't waiting for the MeshGenerators I wouldn't have noticed these other failures. Some people do not update their code that frequently to see the deprecation messages either. Perhaps an email to the mailing list when such a fundamental change to the API is changed?

I know we regularly check the CIVET pages and it would be nice if our application developers did as well, but an email could be a good thing. Most of our users will have no idea what we are talking about, but that's probably ok.

> I am fine with making the necessary API changes to the downstream apps

I really appreciate it. We do check the civet pages regularly, and certainly updating dowstream apps to my new API is on my TO-DO list, but if an application developer gets to it before me, it really makes me happy.
